pkg/events/derive/hidden_kernel_module.go

```diff
@@ -25,69 +26,188 @@ import (
 		if err != nil {
 			return nil, err
 		}
 
-		name, err := parse.ArgVal[[]byte](&event, "name")
+		if _, found := foundHiddenKernModsCache.Get(address); found {
+			return nil, nil // Already reported this event - no need to report again
+		}
+
+		flags, err := parse.ArgVal[uint32](&event, "flags")
 		if err != nil {
 			return nil, err
 		}
 
-		if _, found := foundHiddenKernModsCache.Get(address); found {
-			return nil, nil // already reported this event - no need to report again
-		} else {
-			foundHiddenKernModsCache.Add(address, struct{}{}) // so we won't report multiple times
+		var nameBytes []byte
+		var name string
+
+		if flags&HiddenModule != 0 {
+			// Continue to submit the event to the user!
+		} else if flags&procModules != 0 {
+			// Check again against /proc/modules to address possible race condition
+			nameBytes, err = parse.ArgVal[[]byte](&event, "name")
+			if err != nil {
+				return nil, err
+			}
+			name = string(nameBytes[:bytes.IndexByte(nameBytes[:], 0)]) // Remove the trailing terminating characters
+
+			isFound, err := recheckProcModules(name)
+			if err != nil {
+				return nil, err
+			}
+
+			if isFound {
+				return nil, nil // It's not hidden
+			}
+		} else if flags&SendToVerifier != 0 {
+			// Do another check for this particular module against the modules list once again to tackle race conditions.
+			// If it still isn't found, we'll receive this event once again with flags that will cause it to be sent to user
+			wakeupChannel <- ScanRequest{Address: address, Flags: flags}
+			return nil, nil
+		} else if flags&FullScan != 0 {
+			wakeupChannel <- ScanRequest{Flags: flags} // No need to send the address as we're doing a full generic scan
+			return nil, nil
+		} else if flags&NewMod != 0 {
+			startScanTime := address // The address field in this case is unused, so it's being used as the start scan time
+			err := newModsCheckForHidden(startScanTime, flags)
+			return nil, err
+		}
+
+		if nameBytes == nil { // If we didn't parse it in procModules condition above, parse it now
+			nameBytes, err := parse.ArgVal[[]byte](&event, "name")
+			if err != nil {
+				return nil, err
+			}
+			name = string(nameBytes[:bytes.IndexByte(nameBytes[:], 0)]) // Remove the trailing terminating characters
+		}
+
+		srcversionBytes, err := parse.ArgVal[[]byte](&event, "srcversion")
+		if err != nil {
+			return nil, err
 		}
+		srcversion := string(srcversionBytes[:bytes.IndexByte(srcversionBytes[:], 0)]) // Remove the trailing terminating characters
 
 		addrHex := fmt.Sprintf("0x%x", address)
 		if len(addrHex) == 2 {
 			logger.Warnw("Failed converting module address to hex")
 		}
 
-		return []interface{}{addrHex, string(name)}, nil
+		foundHiddenKernModsCache.Add(address, struct{}{}) // So we won't report multiple times
+
+		return []interface{}{addrHex, name, srcversion}, nil
 	}
 }
 
-func InitFoundHiddenModulesCache() error {
+func newModsCheckForHidden(startScanTime uint64, flags uint32) error {
```

Comment:
Could you please "godoc" all the functions with a brief description of what they do and why ?